crates/common/tedge_utils/src/operation_timer.rs

```diff
@@ -0,0 +1,74 @@
+use std::{collections::HashMap, fmt::Debug, hash::Hash, pin::Pin, time::Duration};
+
+use futures::{stream::FuturesUnordered, Future, StreamExt};
+use tokio::time::{sleep, Instant};
+
+pub const DEFAULT_OPERATION_TIMEOUT: Duration = Duration::from_secs(10); //TODO: Make this configurable?
+
+pub type TimerFuture<T> = Pin<Box<dyn Future<Output = T> + Send + Sync>>;
+
+#[derive(Debug, Default)]
+pub struct OperationTimer<T> {
+    unfinished_child_op_status_map: HashMap<T, (ActiveOperationState, Instant)>,
+    unfinished_child_op_timers: FuturesUnordered<TimerFuture<(T, Instant)>>,
+}
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum ActiveOperationState {
+    Pending,
+    Executing,
+}
+
+impl<T> OperationTimer<T>
+where
+    T: Debug + Eq + Hash + Clone + Send + Sync + 'static,
+{
+    pub fn start_operation_timer(
+        &mut self,
+        operation_id: T,
+        operation_state: ActiveOperationState,
+    ) {
+        let timer_start = Instant::now();
+        let op_timer = Box::pin(Self::timer_for_child_op(operation_id.clone(), timer_start));
+        self.unfinished_child_op_timers.push(op_timer);
+        self.unfinished_child_op_status_map
+            .insert(operation_id, (operation_state, timer_start));
+    }
+
+    pub fn stop_operation_timer(&mut self, operation_id: T) {
+        self.unfinished_child_op_status_map.remove(&operation_id);
+    }
+
+    pub fn current_operation_state(&self, operation_id: &T) -> Option<ActiveOperationState> {
+        self.unfinished_child_op_status_map
+            .get(operation_id)
+            .map(|value| value.0)
+    }
+
+    pub async fn next_timed_out_operation(&mut self) -> Option<(T, ActiveOperationState)> {
```

Comment:
Such a complicated logic is required to find the next timed-out operation because `FuturesUnordered` doesn't let you cancel/reset a `TimerFuture` that was already added to it. So, a timer once scheduled (while triggering the operation request) will definitely time-out eventually, even if that operation completes, or tries to reset the timer with an "executing" status. So, the idea is to start "new" timers on receipt of every "executing" response and when a timer times-out, make sure that it was the latest timer started for that operation.